lib/elixir/src/elixir_with.erl

```diff
@@ -70,37 +70,49 @@ translate(Meta, Args, S) ->
   {Parts, [{do, Expr} | ExprList]} = elixir_utils:split_last(Args),
   case ExprList of
     [{else, ElseExpr}] ->
-      {TCases, TS} = translate_case(Parts, {ok, Expr}, fun(X) -> {error, X} end, S),
-      translate_else(Meta, TCases, ElseExpr, TS);
+      {TCase, TS, HasMatch} = translate_case(Parts, {ok, Expr}, fun(X) -> {error, X} end, S),
+      translate_else(Meta, TCase, ElseExpr, TS, HasMatch);
     [] ->
-      translate_case(Parts, Expr, fun(X) -> X end, S)
+      {TCase, TS, _HasMatch} = translate_case(Parts, Expr, fun(X) -> X end, S),
+      {TCase, TS}
   end.
 
 translate_case(Parts, DoExpr, Wrapper, S) ->
-  Cases = build_case(Parts, DoExpr, Wrapper),
-  {TCases, TS} = elixir_translator:translate(Cases, S#elixir_scope{extra=nil}),
-  {TCases, elixir_scope:mergec(S, TS)}.
-
-translate_else(Meta, WithCases, ElseExpr, S) ->
+  {Case, HasMatch} = build_case(Parts, DoExpr, Wrapper, false),
+  {TCase, TS} = elixir_translator:translate(Case, S#elixir_scope{extra=nil}),
+  {TCase, elixir_scope:mergec(S, TS), HasMatch}.
+
+translate_else(Meta, WithCases, _ElseExpr, S, false) ->
+  Warning = "else clauses never match",
```

Comment:
Okay, that makes sense.

On Friday, 2 September 2016, Andrea Leopardi <notifications@github.com>
wrote:

> In lib/elixir/src/elixir_with.erl
> <https://github.com/elixir-lang/elixir/pull/5179#discussion_r77362831>:
>
> >    end.
> >
> >  translate_case(Parts, DoExpr, Wrapper, S) ->
> > -  Cases = build_case(Parts, DoExpr, Wrapper),
> > -  {TCases, TS} = elixir_translator:translate(Cases, S#elixir_scope{extra=nil}),
> > -  {TCases, elixir_scope:mergec(S, TS)}.
> > -
> > -translate_else(Meta, WithCases, ElseExpr, S) ->
> > +  {Case, HasMatch} = build_case(Parts, DoExpr, Wrapper, false),
> > +  {TCase, TS} = elixir_translator:translate(Case, S#elixir_scope{extra=nil}),
> > +  {TCase, elixir_scope:mergec(S, TS), HasMatch}.
> > +
> > +translate_else(Meta, WithCases, _ElseExpr, S, false) ->
> > +  Warning = "else clauses never match",
>
> @ericmj <https://github.com/ericmj> I don't think so, if you have even
> just one pattern in the with that may not match, then the else clause is
> valid to be there and can be used. The problem is that if all patterns in
> the with always match, then we'll never get in the else. Does that make
> sense?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/5179/files/bb296927711faaaa6210550cf174bf62d2745362#r77362831>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATV2hmkR1eTj4oKDTXfDUWBxhdUn39-ks5qmD0wgaJpZM4JyJG7>
> .
>


-- 
Eric Meadows-Jönsson